### PR TITLE
Result Events

### PR DESCRIPTION
**Pull request by clebergnu**
This is a complete implementation of the `ResultEvents` plugin interface, including the migration of the `HumanResult`, `TAP` and `Journal` plugins.  A few notes:

 * The `FileOrStdOutAction` approach was kept to lock access to STDOUT.  As suggested by @ldoktor there are fancier and cleaner approaches, but they were not attempted here.

 * Those changes expose a lot of issues about or current job/runner implementation details.  Some of the commits are controversial and inteded as a RFC.

### CODE REVIEW

**Comment by clebergnu** [comment]
@apahim @ldoktor looks like travis is having some issues (with this PR only!?).  I've gone through each commit in the series and run `make check` on each one and can't reproduce the failures.

I've resent the build, in case it was a temporary issue.

Anyway, if you guys have time, please take a look and test this PR.

Thanks!

**Comment by clebergnu** [comment]
This has passed on Python 2.7, and the failure on Python 2.6 seems to be a timing related issue at `test_kill_stopped_sleep (selftests.functional.test_basic.RunnerSimpleTest)`.

**Comment by ldoktor** [comment]
The `test_kill_stopped_sleep` is quite sensitive to timing and I also saw false-negatives. Anyway now it looks like it passed...

File: avocado/plugins/run.py [hunk 1]
```diff
@@ -162,9 +162,11 @@ def run(self, args):
             sys.exit(exit_codes.AVOCADO_FAIL)
         job_instance = job.Job(args)
         job_run = job_instance.run()
-        result_dispatcher = ResultDispatcher()
-        if result_dispatcher.extensions:
-            result_dispatcher.map_method('render',
-                                         job_instance.result,
-                                         job_instance)
+
+        if job_instance.status != 'RUNNING':
```
Comment (ldoktor) [inline avocado/plugins/run.py]: I think the output plugins should be always called. They receive the job_instance, therefor they can query for the status and treat it (eg. by storing the error or original status)

Comment (ldoktor) [inline avocado/plugins/run.py]: (Consider an improved human UI plugin. You do want to receive all the information, not only when the system finishes properly)

File: avocado/core/result.py [hunk 2]
```diff
@@ -125,7 +125,7 @@ def start_tests(self):
         """
         Called once before any tests are executed.
         """
-        self.tests_run += 1
+        pass
```
Comment (ldoktor) [inline avocado/core/result.py]: I always disliked this, but had no time to investigate it thoroughly. Thanks :smile:

File: avocado/core/job.py [hunk 3]
```diff
@@ -106,7 +106,6 @@ def __init__(self, args=None):
             self.loglevel = logging.DEBUG
 
         self.status = "RUNNING"
-        self.result_proxy = result.ResultProxy()
```
Comment (ldoktor) [inline avocado/core/job.py]: +11/-126, I like those patches :smile:

File: avocado/plugins/run.py (hunk 1 above)
Comment (ldoktor) [inline avocado/plugins/run.py]: OK I played with it and this patch prevents things like this:

```
avocado run fsafdsfa

Unable to resolve reference(s) 'fsafdsfa' with plugins(s) 'file', 'external', try running 'avocado list -V fsafdsfa' to see the details.
JOB HTML   : /home/medic/avocado/job-results/job-2016-11-03T19.54-1a6f1da/html/results.html
```

But I still believe it should be treated inside the plugins as with this patch you have no way to react on it. But if we keep executing the `render` methods even on RUNNING jobs we have the greater flexibility for plugins.

**Comment by ldoktor** [comment]
Overall I like it. I'll play with it more tomorrow, also I'll double-check it fixes all places etc, but the first impression is good. There is just one think which I don't like and that is the order of Plugin code and then the hooks. I think it should be the other way around, because if one checkouts the commit 12053e7 they might be confused why their plugin does not work.

Personally I'd merge the patches into one as they treat the same thing and depend on each other. If you prefer this way you should first add the functionality and then define the plugin interface to avoid confusion.

**Comment by clebergnu** [comment]
Closing in favor of #1595

**Pull request closed without merging**